lib/ecto/repo/queryable.ex

```diff
@@ -124,6 +125,75 @@ defmodule Ecto.Repo.Queryable do
     end
   end
 
+  def insert_all(name, schema, queryable, opts) when is_atom(schema) do
+    do_insert_all(
+      name,
+      schema,
+      schema.__schema__(:prefix),
+      schema.__schema__(:source),
+      queryable,
+      opts
+    )
+  end
+
+  def insert_all(name, table, queryable, opts) when is_binary(table) do
+    do_insert_all(name, nil, nil, table, queryable, opts)
+  end
+
+  def insert_all(name, {source, schema}, queryable, opts) when is_atom(schema) do
+    do_insert_all(name, schema, schema.__schema__(:prefix), source, queryable, opts)
+  end
+
+  defp do_insert_all(name, schema, prefix, source, queryable, opts) when is_list(opts) do
+    query = Ecto.Queryable.to_query(queryable)
+    {adapter, %{cache: cache, repo: repo} = adapter_meta} = Ecto.Repo.Registry.lookup(name)
+    autogen_id = schema && schema.__schema__(:autogenerate_id)
+    dumper = schema && schema.__schema__(:dump)
+
+    {_return_fields_or_types, return_sources} =
+      schema
+      |> Ecto.Repo.Schema.returning(opts)
+      |> Ecto.Repo.Schema.fields_to_sources(dumper)
+
+    schema_meta = Ecto.Repo.Schema.metadata(schema, prefix, source, autogen_id, nil, opts)
+    on_conflict = Keyword.get(opts, :on_conflict, :raise)
+    conflict_target = Keyword.get(opts, :conflict_target, [])
+    conflict_target = Ecto.Repo.Schema.conflict_target(conflict_target, dumper)
+
+    on_conflict =
+      Ecto.Repo.Schema.on_conflict(on_conflict, conflict_target, schema_meta, 0, adapter)
+
+    {query, opts} = prepare_query(:insert_all, repo, query, opts)
+    query = attach_prefix(query, opts)
+    insert_all_tuple = {query, prefix, source, on_conflict, return_sources}
+
+    {query_meta, prepared, params} =
+      Planner.query(insert_all_tuple, :insert_all, cache, adapter, 0)
+
+    case query_meta do
+      %{select: nil} ->
+        adapter.execute(adapter_meta, query_meta, prepared, params, opts)
+
+      %{select: select, sources: sources, preloads: preloads} ->
+        %{
+          preprocess: preprocess,
+          postprocess: postprocess,
+          take: take,
+          assocs: assocs,
+          from: from
+        } = select
+
+        preprocessor = preprocessor(from, preprocess, adapter)
```

Comment:
@josevalim - do I need to call this preprocessor/3 and postprocessor/4 for Queryable.insert_all or should I just call Schema.post_process/5?

Reply:
Hrm... now that I think about it, you are correct. We don't want to call this preprocessor/postprocessor at all. Rather, we want to depend on everything the Ecto.Repo.Schema does about the return type. In fact, we may even want to use the same adapter.insert_all instructions, except that we pass a query instead of `rows` here: https://github.com/elixir-ecto/ecto/pull/3418/files#diff-053c85c1728be8c0624c7e24d6eddc83R54

Given we ended-up invoking a bunch of functions in Ecto.Repo.Schema, I wonder if it would be better to move this functionality to Ecto.Repo.Schema altogether.

The other thing is that we need to guarantee that the select returns a map and the map keys must match the fields in the schema. Those are going to be used as the INTO headers.